cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -621,11 +621,18 @@ class DatasetWriter::DatasetWriterImpl {
         backpressure = writer_state_.open_files_throttle.Acquire(1);
         if (!backpressure.is_finished()) {
           EVENT_ON_CURRENT_SPAN("DatasetWriter::Backpressure::TooManyOpenFiles");
+          writer_state_.rows_in_flight_throttle.Release(next_chunk->num_rows());
```

Comment:
Can we move the `rows_in_flight_throttle.Acquire()` call to close to the existing `rows_in_flight_throttle.Release()` call and this `Release()` call to there? (`DatasetWriterFileQueue`?)
I think that the `Acquire()`/`Release()` pair should be close for easy to maintain like `malloc()`/`free()` pair and `new`/`delete` pair.

(Can we apply the same approach to the `open_files_throttle.Acquire()`/`Release()` pair too?)

Reply:
Hmm I think the best way is using something like a "RAII Guard" here. This might need kind of refactor the throttle type here

Reply:
Find it a bit hard to get related...